### PR TITLE
HUB-766: Validate & format entity id's

### PR DESCRIPTION
**Pull request by AmritSidhu**
https://govukverify.atlassian.net/browse/HUB-766

- Added validation to the new_msa_component_event when the component is being created
- Fixed existing tests relying on the same entity id across all MSA components, now uses SecureRandom.alphanumeric to generate the entity id
- Created a system test to show that you cannot create a second MSA component with an existing entity id
- Also fixed a error display message.

https://govukverify.atlassian.net/browse/HUB-707

- Removed trailing and leading whitespaces from service/msa entity_id
- Raises an error if there are spaces between the string
- Created system / model test to prove this works

The new validation displays: 
Screenshot 2021-01-21 at 11 48 17

Screenshot 2021-01-21 at 11 48 54

### CODE REVIEW

**Review by instantiator — APPROVED** [review]

File: spec/models/publish_services_event_metadata_spec.rb [hunk 1]
```diff
@@ -340,7 +340,7 @@
             .times(3)
 
           expect(Rails.logger).to receive(:error)
-            .with("Error getting signing certificates for entity_id: #{new_msa_encryption_certificate.component.entity_id}! (Code: 404)")
+            .with("Error getting signing certificates for entity_id: #{msa_signing_certificate.component.entity_id}! (Code: 404)")
```
Comment (CharlesIC) [inline spec/models/publish_services_event_metadata_spec.rb:343]: Why has this changed?

File: spec/system/visit_msa_component_new_spec.rb [hunk 2]
```diff
@@ -40,6 +41,22 @@
     end
   end
 
+  context 'creation fails without unique entity id' do
+    it 'when entity id already exists for a msa component' do
+      existing_component = msa_component
+      entity = 'existing entity_id'
```
Comment (CharlesIC) [inline spec/system/visit_msa_component_new_spec.rb:47]: I believe entity IDs shouldn't be allowed to have spaces? This should also be validated (not necessarily part of this story but maybe worth fixing whilst we're at it)

File: spec/models/publish_services_event_metadata_spec.rb (hunk 1 above)
Comment (AmritSidhu) [reply spec/models/publish_services_event_metadata_spec.rb:343]: So there are two identical set of tests, one for encryption, one for signing. I assume it was copy and pasted over and someone didnt change it, so it has always been checking the wrong thing.

**Comment by CharlesIC** [comment]
There are some test failures in Travis

```
An error occurred while loading ./spec/system/visit_msa_component_new_spec.rb.
Failure/Error: visit new_msa_component_path
NameError:
  undefined local variable or method `new_msa_component_path' for #<Class:0x0000000005e21f80>
```

**Comment by instantiator** [comment]
This looks good to me, and the tests pass. I'm still figuring out how I can run a self-service instance locally to test - but happy to give it a check in return for a quick demo on Monday.

**Review by instantiator — DISMISSED** [review]

File: spec/system/visit_msa_component_new_spec.rb (hunk 2 above)
Comment (CharlesIC) [inline spec/system/visit_msa_component_new_spec.rb]: Wouldn't this fail now due to your new validation to not have spaces in entity ids?

File: spec/models/new_msa_component_event_spec.rb [hunk 3]
```diff
@@ -31,4 +32,18 @@
...
+
+  context 'entity_id' do
+    it 'must be provided' do
+      event = build(:new_msa_component_event, entity_id: '')
+      expect(event).to_not be_valid
+      expect(event.errors[:entity_id]).to eql [t('components.errors.missing_entity_id')]
+    end
+
+    it 'must not contain spaces' do
+      event = build(:new_msa_component_event, entity_id: ' test entity id ')
```
Comment (CharlesIC) [inline spec/models/new_msa_component_event_spec.rb:44]: This is mixing two checks in one

This should be split into two tests:
- must not contain spaces - provide test entity id 'test entity id' (this test)
- strips whitespace - provide test entity id ` test-entity-id ` and check the created component has no spaces in entity id

File: spec/system/visit_msa_component_new_spec.rb [hunk 4]
```diff
@@ -53,7 +53,22 @@
...
       click_button 'Create MSA component'
 
-      expect(page).to have_content t('components.errors.invalid_entity_id')
+      expect(page).to have_content t('components.errors.existing_entity_id')
+    end
+  end
+
+  context 'creation fails with whitespaces within entity id' do
+    it 'when entity id in a msa component has spaces' do
+      entity = ' test entity id '
```
Comment (CharlesIC) [inline spec/system/visit_msa_component_new_spec.rb:62]: Same comment here - test stripping and failing on whitespace in the middle in two separate tests

File: spec/system/visit_service_new_spec.rb [hunk 5]
```diff
@@ -42,7 +42,18 @@
       fill_in 'service_entity_id', with: entity_id
       click_button 'Create service'
 
-      expect(page).to have_content 'Entity ID is required'
+      expect(page).to have_content I18n.t('service.errors.missing_entity_id')
+    end
+
+    it 'when entity id has spaces between words' do
+      service_name = 'test component'
```
Comment (CharlesIC) [inline spec/system/visit_service_new_spec.rb:49]: Same comment here - test stripping and failing on whitespace in the middle in two separate tests

**Review by CharlesIC — COMMENTED** [review]
This looks great - just a small suggestion regarding splitting the tests to be separate

**Review by instantiator — APPROVED** [review]

File: spec/models/new_msa_component_event_spec.rb [hunk 6]
```diff
@@ -31,4 +32,24 @@
...
+      event = build(:new_msa_component_event, entity_id: '')
+      expect(event).to_not be_valid
+      expect(event.errors[:entity_id]).to eql [t('components.errors.missing_entity_id')]
+    end
+
+    it 'is valid when there is only leading and trailing whitespaces' do
+      event = build(:new_msa_component_event, entity_id: ' https://test-entity-id ')
+      expect(event).to be_valid
+      expect(event.errors[:entity_id]).to eql []
+    end
```
Comment (CharlesIC) [inline spec/models/new_msa_component_event_spec.rb:47]: Can we also check that the whitespaces are stripped? Does the `event` object contain an entity ID string of `'https://test-entity-id'`. This test correctly checks that leading and trailing whitespaces don't fail validation but it doesn't verify that we actually strip them

File: spec/system/visit_msa_component_new_spec.rb [hunk 7]
```diff
@@ -40,6 +41,52 @@
...
+      select team.name, from: "component_team_id"
+      choose('component_environment_staging')
+      click_button 'Create MSA component'
+
+      expect(page).to have_content t('components.errors.existing_entity_id')
+    end
+  end
+
+  context 'creation is successful with only leading and trailing whitespaces on entity id' do
+    it 'when entity id in a msa component has leading and trailing whitespaces' do
```
Comment (CharlesIC) [inline spec/system/visit_msa_component_new_spec.rb:61]: Can this test validate that the entity ID of the newly created component is `'http://test-entity-id-2'`? I.e. verify that whitespaces have been stripped

File: spec/system/visit_service_new_spec.rb [hunk 8]
```diff
@@ -18,6 +18,17 @@
 
       expect(current_path).to eql admin_path
     end
+
+    it 'when entity id has only leading and trailing spaces' do
```
Comment (CharlesIC) [inline spec/system/visit_service_new_spec.rb:22]: Same comment here about testing - can we check the leading and trailing whitespaces have been removed?

**Review by instantiator — APPROVED** [review]

**Review by j-maynard — APPROVED** [review]
🦁

**Pull request merged**